ulib/LowStar.Vector.fst

```diff
@@ -0,0 +1,558 @@
+module LowStar.Vector
+
+open FStar.All
+open FStar.Integers
+open LowStar.Modifies
+
+module HH = FStar.Monotonic.HyperHeap
+module HS = FStar.HyperStack
+module HST = FStar.HyperStack.ST
+module MHS = FStar.Monotonic.HyperStack
+module B = LowStar.Buffer
+module S = FStar.Seq
+
+type uint32_t = UInt32.t
+
+val max_uint32: uint32_t
+let max_uint32 = UInt32.uint_to_t (UInt.max_int UInt32.n)
+
+module U32 = FStar.UInt32
+
+/// Some helpers for `LowStar.Buffer`
+
+val get_as_seq_index:
+  #a:Type -> h:HS.mem -> buf:B.buffer a -> i:uint32_t{i < B.len buf} ->
+  Lemma (B.get h buf (U32.v i) == S.index (B.as_seq h (B.gsub buf i 1ul)) 0)
+let get_as_seq_index #a h buf i = ()
+
+val get_preserved_buffer:
+  #a:Type -> buf:B.buffer a ->
+  i:uint32_t -> len:uint32_t{U32.v i + U32.v len <= B.length buf} ->
+  k:uint32_t{(k < i || i + len <= k) && k < B.len buf} ->
+  h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (B.live h0 buf /\
+		  modifies (B.loc_buffer (B.gsub buf i len)) h0 h1))
+	(ensures (B.get h0 buf (U32.v k) == B.get h1 buf (U32.v k)))
+let get_preserved_buffer #a buf i len k h0 h1 =
+  get_as_seq_index h0 buf k; get_as_seq_index h1 buf k
+
+/// Abstract vector type
+
+noeq type vector_str a =
```

Comment:
You might want to add `inline_for_extraction` on the type too, if you're hoping to see the projectors for the type inlined, too.